### PR TITLE
Game refactor

### PR DESCRIPTION
**Pull request by RicsiToth**

### CODE REVIEW

File: src/model/Expression.js [hunk 1]
```diff
@@ -18,6 +18,9 @@ class Expression {
     return null;
   }
 
+  getVariables(){
+    return null;
+  }
```
Comment (crnkjck) [inline src/model/Expression.js:23]: Do not return `null`, if you can return a sane type-safe default – `[]` in this case.

File: src/model/formula/Formula.Conjunction.js [hunk 2]
```diff
@@ -58,6 +58,16 @@ class Conjunction extends Formula {
...
   }
+
+  getSubFormulasCommitment(commitment){
+    return [commitment, commitment];
+  }
+
+  getVariables(){
+    const leftVariables = this.subLeft.getVariables();
+    const rightVariables = this.subRight.getVariables()
+    return leftVariables.concat(rightVariables);
```
Comment (crnkjck) [inline src/model/formula/Formula.Conjunction.js:69]: This is a one-liner:
```javascript
getVariables: () =>
    this.subLeft.getVariables().concat(this.subRight.getVariables());
```

File: src/model/formula/Formula.Disjunction.js [hunk 3]
```diff
@@ -58,6 +58,16 @@ class Disjunction extends Formula {
     this.subLeft.setVariable(from, to);
     this.subRight.setVariable(from, to);
   }
```
Comment (crnkjck) [inline src/model/formula/Formula.Disjunction.js:60]: `setVariable` is a confusing name for a very standard operation – substitution.

So call this `substitute` and allow any term to be substituted for the variable, not just another variable.

It’s also bad practice to modify the original formula by substitution. Make a new copy on the fly, e.g., in this case:
```javascript
substitute: (var, term) =>
  new Disjunction(
    this.subLeft.substitute(var, term),
    this.subRight.substitute(var, term)
  );
```
If all `substitute` methods create new copies, then this is safe and will create a completely new formula.

Make sure that also constants and quantified formulas that quantify `var` return a new copy of themselves in `substitute`.

File: src/model/formula/Formula.Equivalence.js [hunk 4]
```diff
@@ -47,17 +51,29 @@ class Equivalence extends Formula {
...
 
     getType(commitment){
-        return commitment ? GAME_EQUIVALENCE : PLAYER_EQUIVALENCE;
+        return commitment ? GAME_OPERATOR : PLAYER_OPERATOR;
     }
 
     getSubFormulas(){
-        return [this.subLeft, this.subRight];
+        let toRightImpl = new Implication(this.subLeft, this.subRight);
+        let toLeftImpl = new Implication(this.subRight, this.subLeft);
```
Comment (crnkjck) [inline src/model/formula/Formula.Equivalence.js:59]: `const`

File: src/model/formula/Formula.Implication.js [hunk 5]
```diff
@@ -58,6 +58,19 @@ class Implication extends Formula {
     this.subLeft.setVariable(from, to);
     this.subRight.setVariable(from, to);
   }
+
+  getSubFormulasCommitment(commitment){
+    if(commitment){
+      return [false, true];
+    }
+    return [true, false];
```
Comment (crnkjck) [inline src/model/formula/Formula.Implication.js:66]:
```js
return [!commitment, commitment];
```

File: src/model/term/Term.FunctionTerm.js [hunk 6]
```diff
@@ -64,6 +64,10 @@ class FunctionTerm extends Term {
   setVariable(from, to){
     this.terms.forEach(term => term.setVariable(from, to));
   }
+
+  getVariables(){
+    return this.terms.map(term => term.getVariables()).flat();
```
Comment (crnkjck) [inline src/model/term/Term.FunctionTerm.js:69]: [`flatMap(… => …)`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap) is more efficient than `map(… => …).flat()`

File: src/model/formula/Formula.PredicateAtom.js [hunk 7]
```diff
@@ -75,6 +75,14 @@ class PredicateAtom extends Formula {
   setVariable(from, to){
     this.terms.forEach(term => term.setVariable(from, to));
   }
+
+  getSubFormulasCommitment(commitment){
+    return [];
+  }
+
+  getVariables(){
+    return this.terms.map(term => term.getVariables()).flat();
```
Comment (crnkjck) [inline src/model/formula/Formula.PredicateAtom.js:84]: [`flatMap(… => …)`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap) is more efficient than `map(… => …).flat()`

**Review by crnkjck — COMMENTED** [review]
A quick `model` review.

File: src/model/formula/Formula.Disjunction.js (hunk 3 above)
Comment (RicsiToth) [reply src/model/formula/Formula.Disjunction.js:60]: I don't understand what you are trying to say here: "So call this substitute and allow any term to be substituted for the variable, not just another variable.". Are you implying that Constants should be substituted as well?

**Pull request merged**